### PR TITLE
chore(Coder plugin): Create guide for working with the Coder SDK

### LINKED ISSUE #112: Coder plugin: Add documentation for helping users cache state from `CoderClient` API calls
Part of umbrella issue #16. Depends on final implementations of  #107 and #114.

## Requirements
- We add documentation for some of the various ways that a user can turn the data returned by `CoderClient` into React state. We also detail the pros/cons of each approach
   - `useEffect` + `useState`
   - `useAsync` (go-to Backstage recommendation, but usually underpowered for more sophisticated UI)
   - `useQuery`

### PR DESCRIPTION
**Pull request by Parkreiner**
Closes #112

## Changes made
- Added guide

## Things left to do
- [ ] Add images/video to designated areas of the draft
- [ ] Add link to the guide from the main README

## Notes
- Might be helpful to view the README [in the actual repo view](https://github.com/coder/backstage-plugins/blob/mes/vendored-sdk-readme/plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md)
- I updated a couple of the other pending PRs as I was writing this out
   - There's a couple of things I'm realizing could be improved with the final design, but I'm going to make a note of them, and come back once we have more customer interest

### CODE REVIEW

File: plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md [hunk 1]
```diff
@@ -0,0 +1,557 @@
+# Using the Coder SDK
+
+This document walks you through using the Coder SDK from within Backstage.
```
Comment (Kira-Pilot) [inline plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md:3]: I wonder if the usage of `SDK` here might feel misleading. What do you think about rephrasing to `using the Coder API` or something similar?

File: plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md [hunk 2]
```diff
@@ -0,0 +1,557 @@
...
+
+## Disclaimer
+
+As of May 2024, Coder does not have a fully public, versioned SDK published on a package manager like NPM. Coder does intend to release a true JavaScript/TypeScript SDK, but until that is released, the version exposed through Backstage can be best thought of as a "preview"/"testbed" SDK. It is also used in production for [the official Coder VS Code extension](https://github.com/coder/vscode-coder).
```
Comment (Kira-Pilot) [inline plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md:7]: 🤔 Hmmm, I wonder if we feel comfortable putting this on our roadmap.

@bpmct What do you think?

File: plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md [hunk 3]
```diff
@@ -0,0 +1,557 @@
...
+  - Accessing the SDK via `useCoderSdk`
+  - Accessing the SDK via `useCoderQuery`
+- Authenticating
+  - The fallback auth UI
+- Performing queries
+  - Recommendations for caching data
+  - How the Coder plugin caches data
+  - The `useCoderQuery` custom hook
+  - Cautions against other common UI caching strategies
+- Performing mutations
```
Comment (Kira-Pilot) [inline plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md:29]: Do we want to create some inner-document links here?

File: plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md [hunk 4]
```diff
@@ -0,0 +1,557 @@
...
+
+### Before you begin
+
+This guide assumes that you have already added the `CoderProvider` component to your Backstage deployment. If you have not, [please see the main README](../../README.md#setup) for instructions on getting that set up.
+
+## Quick-start
+
+### The main hooks to use when working with the SDK
+
+There are about six main hooks that you want to consider when using the Coder SDK.These can be broken down into two main categories:
```
Comment (Kira-Pilot) [inline plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md:39]:
```suggestion
There are six main hooks that you want to consider when using the Coder SDK.These can be broken down into two main categories:
```

File: plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md [hunk 5]
```diff
@@ -0,0 +1,557 @@
...
+  - Cautions against other common UI caching strategies
+- Performing mutations
+
+### Before you begin
+
+This guide assumes that you have already added the `CoderProvider` component to your Backstage deployment. If you have not, [please see the main README](../../README.md#setup) for instructions on getting that set up.
+
+## Quick-start
+
+### The main hooks to use when working with the SDK
```
Comment (Kira-Pilot) [inline plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md:37]:
```suggestion
### Hook overview
```

File: plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md [hunk 6]
```diff
@@ -0,0 +1,557 @@
...
+- `useCoderQuery`
+
+### Accessing the SDK through `useCoderSdk`
+
+`useCoderSdk` is a lower-level "primitive" for accessing the Coder SDK. The SDK exposes a mix of different REST API methods for interacting with your Coder deployment's resources.
+
+Calling the hook will give you an object with all available API methods. As these methods are all async, **none** of them are suitable for use in render logic. They must be called from within effects or event handlers.
+
+```tsx
+// Illustrative example - these patterns are a very bad idea in production!
```
Comment (Kira-Pilot) [inline plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md:67]: I get why you put this warning here but combined with the async reminder in the paragraph above, it took me a second to understand you were exemplifying usage (rather than including a 'don't do this' snippet). Maybe we could consider rewording slightly?

File: plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md [hunk 7]
```diff
@@ -0,0 +1,557 @@
...
+
+## Authentication
+
+All API methods from the SDK will throw an error if the user is not authenticated. The Coder plugin provides a few different ways of letting the user authenticate with Coder:
+
+- The official Coder components
+- The `CoderProvider` component's fallback auth UI
+- The `useCoderAuth` hook
+
+All three solutions, directly or indirectly, involve the `CoderAuth` type. More information can be found under the `useCoderAuth` section.
```
Comment (Kira-Pilot) [inline plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md:157]: Should we add a document link here

File: plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md [hunk 8]
```diff
@@ -0,0 +1,557 @@
...
+Every official Coder component (such as `CoderWorkspacesCard`) exported through the plugin is guaranteed to have some mechanism for supplying auth information. This is typically done via a UI form.
+
+<-- [Include screenshot/video of auth input here] -->
+
+#### Pros
+
+- Come pre-wired with the ability to let the user supply auth information
+- Extensively tested
+- Let you override the styles or re-compose the individual pieces for your needs
+- Always audited for WCAG Level AA accessibility, and include landmark behavior for screen readers
```
Comment (Kira-Pilot) [inline plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md:170]:
```suggestion
- Always audited for WCAG Level AA accessibility, and includes landmark behavior for screen readers
```

File: plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md [hunk 9]
```diff
@@ -0,0 +1,557 @@
...
+
+When you include the `CoderProvider` component in your Backstage deployment, you have the option to set the value of `fallbackAuthUiMode`. This value affects how `CoderProvider` will inject a fallback auth input into the Backstage deployment's HTML. This means that, even if you don't use any Coder components, or are on a page that can't use them, users will always have some way of supplying auth information.
+
+```tsx
+<CoderProvider appConfig={youCustomAppConfig} fallbackAuthUiMode="assertive">
+  <RestOfYourComponentsThatUseTheCoderApis />
+</CoderProvider>
+```
+
+<-- [Include screenshot/video of fallback auth input here] -->
```
Comment (Kira-Pilot) [inline plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md:189]: Another note so that we remember to put this in

File: plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md [hunk 10]
```diff
@@ -0,0 +1,557 @@
...
+
+The fallback auth UI will never be visible while the user is authenticated. However, if the user is **not** authenticated, then the value of `fallbackAuthUiMode` will affect what appears on screen:
+
+- `restrained` (default) - The fallback auth UI will not appear if there are official Coder components on screen.
+- `hidden` - The fallback auth is **never** visible on the page. If no official Coder components are on screen, you will need to import `useCoderAuth` into your custom components to authenticate your users.
+- `assertive` - The fallback auth UI is always visible when the user is not authenticated, regardless of whether there are any official Coder components on screen.
+
+#### Pros
```
Comment (Kira-Pilot) [inline plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md:197]: I am struggling to find the information I care most about in this README so that I can get up and moving quickly. Do you think we could find some ways to shorten it for clarity? One idea is that was could fold important callouts from the `Pros` and `Cons` headers in each section into the section body itself.

File: plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md [hunk 11]
```diff
@@ -0,0 +1,557 @@
...
+
+- Simplifies how API data is shared throughout an application
+- Manages race conditions, canceling requests, and revalidating requests
+- Provides a shared API for both queries and mutations
+
+### Problems with fetching via `useEffect`
+
+All functions returned by the Coder SDK maintain stable memory references for the entire lifetime of the SDK. In that sense, every one of these functions can be safely placed inside a `useEffect` dependency array to perform data fetching. In theory, `usEffect` can be used to trigger API calls, while the results (and their relevant loading/error/success states) can be stored via `useState`.
+
+In practice, however, this setup causes a lot of problems:
```
Comment (Kira-Pilot) [inline plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md:264]: I think we may be over-explaining React internals here which is leading to a lengthier document. What do you think?

File: plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md (hunk 1 above)
Comment (bpmct) [reply plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md:3]: Agreed, given we don't publish a Javascript SDK this may be best. We only publish a golang SDK 👍🏼 

We should definitely change this once we publish our SDK

File: plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md (hunk 2 above)
Comment (bpmct) [reply plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md:7]: I'm OK listing it as a roadmap item, let's link to the GitHub issue though: https://github.com/coder/coder/issues/13458

@Parkreiner - feel free to rewrite my issue there too. I did something quickly ^

File: plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md [hunk 12]
```diff
@@ -0,0 +1,557 @@
...
+
+### Accessing the SDK through `useCoderSdk`
+
+`useCoderSdk` is a lower-level "primitive" for accessing the Coder SDK. The SDK exposes a mix of different REST API methods for interacting with your Coder deployment's resources.
+
+Calling the hook will give you an object with all available API methods. As these methods are all async, **none** of them are suitable for use in render logic. They must be called from within effects or event handlers.
+
+```tsx
+// Illustrative example - these patterns are a very bad idea in production!
+function ExampleComponent() {
```
Comment (bpmct) [inline plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md:68]: How does this component perform if the user has not linked their Coder account? IMO all of our examples components need to handle this case

File: plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md [hunk 13]
```diff
@@ -0,0 +1,557 @@
...
+    <>
+      <button
+        // The SDK can also be called from any event handler
+        onClick={async () => {
+          const newWorkspace = await sdk.createWorkspace(
+            'organizationId',
+            'userId',
+            {
+              // Properties for making new workspace go here
+            },
```
Comment (bpmct) [inline plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md:98]: Can we fill this with properties so it works out of the box? Or explain to the user where they can find the properties they need to fill? As a non-Typescript expert, I may get stuck here.

File: plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md [hunk 14]
```diff
@@ -0,0 +1,557 @@
...
+- Automatic integration with official Coder components. The auth fallback UI can detect Coder components without you needing to rewrite any code.
+- All auth UI logic has been tested and audited for accessibility at the same standards as the other Coder components
+
+#### Cons
+
+- Even with three options for setting behavior, fallback auth input may not be visible exactly when you want it to be
+- The `restrained` behavior is only effective on pages where you can place official Coder components. If you are not on one of these pages, the fallback auth UI will always be visible until the user authenticates.
+- Fewer options for customizing the styling
+
+### Authenticating via `useCoderAuth`
```
Comment (bpmct) [inline plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md:210]: Can you show an example of how a component works with this?

File: plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md (hunk 10 above)
Comment (bpmct) [reply plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md:197]: Alternatively, we could make a table and link for all 3 with pros and cons to an example component for each (maybe even just linking to the test case is enough)? Agree that someone should be able to weigh pros and cons at a glance and also see an example for each so that can see what fits into what they are trying to achieve.

File: plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md (hunk 11 above)
Comment (bpmct) [reply plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md:264]: I agree. IMO this guide should just provide solid, out-of-the-box example components and snippets that people can learn from. If we want to provide break-glass examples, docs, references, trade-offs, and recommendations we should do it via a separate doc. We want to make this process look extremely simple and easy to do without the plugin author to make a bunch of decisions.

Most people I assume will copy & paste our examples and modify it for their case and not decide to do their own querying and stuff. If they break glass, I like us guiding them to be a better programmer but think it definitely needs to be separated.

**Review by bpmct — APPROVED** [review]
Exciting to see this coming along! I think this guide is great, but let's try to make it a bit more shorter and prescriptive. As a plugin user, I'd love to be creating (or copy & pasting) example components immediately after (or during) reading this.

Right now, I think readers will be a bit intimidated as it makes it look like they need to make a bunch of decisions. I like that we support this, but would like to challenge you to get more prescriptive in this guide and optimize for example components or functions that people can copy & paste for several scenarios:

- custom functions in Skaffolder (create workspace)
- custom components (template list)
- a "break glass" complex thing with less of our magic (idk what)

File: plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md (hunk 1 above)
Comment (code-asher) [reply plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md:3]: I like `using the Coder API` or `making calls to the Coder API`.

File: plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md [hunk 15]
```diff
@@ -0,0 +1,557 @@
...
+- `upsert`
+- `delete`
+
+Depending on the Coder resource, there may be different API methods that work with a single resource vs all resources (e.g., `sdk.getWorkspace` vs `sdk.getWorkspaces`).
+
+Note that all of these functions will throw an error if the user is not authenticated.
+
+### Error behavior
+
+All SDK functions will throw in the event of an error. You will need to provide additional error handling to expose errors as values within the UI. (Tanstack Query does this automatically for you.)
```
Comment (code-asher) [inline plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md:134]: A nit, but I feel like `(Tanstack Query does this automatically for you.)` can be omitted since someone using it probably knows that.  Unless this is a subtle way to try making people switch heh :laughing:

File: plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md [hunk 16]
```diff
@@ -0,0 +1,557 @@
...
+  - How the Coder plugin caches data
+  - The `useCoderQuery` custom hook
+  - Cautions against other common UI caching strategies
+- Performing mutations
+
+### Before you begin
+
+This guide assumes that you have already added the `CoderProvider` component to your Backstage deployment. If you have not, [please see the main README](../../README.md#setup) for instructions on getting that set up.
+
+## Quick-start
```
Comment (code-asher) [inline plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md:35]: For a quickstart I think we might consider just having a prescriptive copy-paste example, maybe something one code block to copy-paste with `useAsync` or whatever the Backstage thing was and one other block for React Query.

I liked the ones you had in the other PR.  Do we need an entire component as an example?  I felt like just seeing the hooks in use was enough for me to understand how to use them in my application.  Just the minimum code required to make an API call.

File: plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md [hunk 17]
```diff
@@ -0,0 +1,557 @@
...
+```tsx
+const workspacesQuery = useCoderQuery({
+  queryKey: ['workspaces'],
+
+  // Access the SDK without needing to import or call useCoderSdk
+  queryFn: ({ sdk }) => sdk.getWorkspaces({ q: 'owner:me' }),
+});
+```
+
+## Authentication
```
Comment (code-asher) [inline plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md:149]: I think we could just be prescriptive about using `CoderProvider` especially since we already made it a prerequisite above, which means we would not need to explain the other types and could omit this section.

Especially once we get oauth, it will not be possible for either us nor the user to provide UI for that, and the fallback UI matches best currently with how oauth will behave, I think (since it also uses a popup), so it might be best not to provide any options that will just have to be removed once oauth is available so transitioning is easy.

To be safest we could not export `useCoderAuth` and change our components to also use the fallback UI (which would no longer be a fallback I guess at that point haha), but out of scope for this PR I think.

File: plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md [hunk 18]
```diff
@@ -0,0 +1,557 @@
...
+
+- Gives you the finest level of control over all auth concerns
+- Easy to import into any component
+
+#### Cons
+
+- Zero UI logic out of the box; you have to make all components yourself
+- Must be called within a `CoderProvider` component
+
+## Caching API data for UIs
```
Comment (code-asher) [inline plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md:238]: I think we could remove this section, it feels out of scope to me.

File: plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md [hunk 19]
```diff
@@ -0,0 +1,557 @@
...
+
+#### Problems retained
+
+- Everything else from the `useEffect` section
+
+#### New problems
+
+- Even though `useAsync`'s API uses dependency arrays, by default, it is not eligible for the exhaustive deps ES Lint rule. This means that unless you update your ESLint rules, you have no safety nets for making sure that your effect logic runs the correct number of times. There are no protections against accidental typos.
+
+## Bring your own Query Client
```
Comment (code-asher) [inline plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md:293]: This feels out of scope to me as well.

File: plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md (hunk 1 above)
Comment (code-asher) [reply plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md:3]: Also if we rename, the filename could be `coder-api.md` or something like that?

File: plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md [hunk 20]
```diff
@@ -0,0 +1,557 @@
...
+    {/*
+     * Exact same component definition, but the location completely changes the
+     * client that gets accessed
+     */}
+    <YourCustomComponent />
+  </CoderProvider>
+</QueryClient>;
+```
+
+## Sharing data between different queries and mutations
```
Comment (code-asher) [inline plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md:340]: For the same reasons, I would say we should omit this at least from this guide.  Our examples should be prescriptive and do the best thing, and if we do end up needing to document advanced use cases we could revisit.

File: plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md [hunk 21]
```diff
@@ -0,0 +1,557 @@
...
+
+Internally, the official components for the Coder plugin use one shared query key prefix value (`CODER_QUERY_KEY_PREFIX`) for every single query and mutation. This ensures that all Coder-based queries share a common root, and can be easily vacated from the query cache in response to certain events like the user unlinking their Coder account (basically "logging out").
+
+This same prefix is exported to users of the plugin. When using the Coder SDK with Tanstack Query, it is **strongly, strongly** recommended that you use this same query prefix. It lets custom components and Coder components share the same underlying data (reducing the total number of requests). But the plugin is also set up to monitor queries with this prefix, so it can automate cache management and has more ways of detecting expired auth tokens.
+
+## Performing queries
```
Comment (code-asher) [inline plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md:346]: I think we could be prescriptive and just recommend `useCoderQuery`, replacing this with a React Query example above.

File: plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md [hunk 22]
```diff
@@ -0,0 +1,557 @@
...
+- Automatically prohibits some deprecated properties from Tanstack Query v4 (using one of them is treated as a type error). This means guaranteed forwards-compatibility with Tanstack Query v5
+
+#### Cons
+
+- Introduces extra overhead to ensure that queries are fully locked down when the user is not authenticated – even if they're not needed for every use case
+- Doesn't offer quite as much fine-grained control compared to `useQuery`
+- Does not simplify process of connecting Coder queries and mutations, because a `useCoderMutation` hook does not exist (yet)
+- `queryKey` is slightly more confusing, because a new value is implicitly appended to the beginning of the array
+
+## Performing mutations
```
Comment (code-asher) [inline plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md:487]: This seems to just be how to do mutations with React Query, I think?  It feels out of scope to me.

**Review by code-asher — COMMENTED** [review]
I only reviewed the Markdown, I think the rest is alyeady committed and only shows because the branch needs to be rebased, right?

I like how thorough this is, but I think it is a little overwhelming.  Two copy-paste examples (React Query and without) and a link to the API docs would be solid, I think.

File: plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md (hunk 3 above)
Comment (Parkreiner) [reply plugins/backstage-plugin-coder/docs/guides/sdk-recommendations.md:29]: Yeah, sorry for not clarifying – definitely planning to add links to everything. Just didn't pay as much attention here, since I wasn't sure if these headers made sense yet

**Comment by Parkreiner** [comment]
@Kira-Pilot Updated the READMEs: there's two main files now:
- [The main guide](https://github.com/coder/backstage-plugins/blob/mes/vendored-sdk-readme/plugins/backstage-plugin-coder/docs/guides/coder-api.md), which is structured as a quick-start
- [An advanced guide](https://github.com/coder/backstage-plugins/blob/mes/vendored-sdk-readme/plugins/backstage-plugin-coder/docs/guides/coder-api-advanced.md), which will be a catch-all for anything that requires more customization

There are a lot more files changed compared to before, but they're all basically file/property renames/reorganization to make things more clear

Can you let me know if the new READMEs feel a bit more scrutible?

File: plugins/backstage-plugin-coder/docs/README.md [hunk 23]
```diff
@@ -2,10 +2,17 @@
 
 For users who need more information about how to extend and modify the Coder plugin. For general setup, please see our main [README](../README.md).
 
-All documentation reflects version `v0.2.0` of the plugin. Note that breaking API changes may continue to happen for minor versions until the plugin reaches version `v1.0.0`.
+All documentation reflects version `v0.2.1` of the plugin. Note that breaking API changes may continue to happen for minor versions until the plugin reaches version `v1.0.0`.
```
Comment (Kira-Pilot) [inline plugins/backstage-plugin-coder/docs/README.md:5]: I think we're basically explaining SemVer here and can leave this out.

File: plugins/backstage-plugin-coder/docs/guides/coder-api.md [hunk 24]
```diff
@@ -0,0 +1,257 @@
...
+#### Convenience hooks
+
+- `useCoderQuery` - Simplifies wiring up `useQuery`, `useCoderApi`, and `useCoderAuth`
+
+## Recommendations for accessing the API
+
+1. If querying data, prefer `useCoderQuery`. It automatically wires up all auth logic to React Query (which includes pausing queries if the user is not authenticated). It also lets you access the Coder API via its query function. `useQuery` is also a good escape hatch if `useCoderQuery` doesn't meet your needs, but it requires more work to wire up correctly.
+2. If mutating data, you will need to call `useMutation`, `useQueryClient`, and `useCoderApi` in tandem\*. The plugin exposes a `CODER_QUERY_KEY_PREFIX` constant that you can use to group all Coder queries together.
+
+We highly recommend **not** fetching with `useState` + `useEffect`, or with `useAsync`. Both face performance issues when trying to share state. See [ui.dev](https://www.ui.dev/)'s wonderful [_The Story of React Query_ video](https://www.youtube.com/watch?v=OrliU0e09io) for more info on some of the problems they face.
```
Comment (Kira-Pilot) [inline plugins/backstage-plugin-coder/docs/guides/coder-api.md:43]: Any reason we have to include this disclaimer? Seems like general React knowledge, rather than Coder plugin specific knowledge.

File: plugins/backstage-plugin-coder/docs/guides/coder-api.md [hunk 25]
```diff
@@ -0,0 +1,257 @@
...
+| Can retain state after component unmounts                          | 🚫                       | 🚫         | ✅         | ✅              |
+| Easy, on-command query invalidation                                | 🚫                       | 🚫         | ✅         | ✅              |
+| Automatic retry logic when a query fails                           | 🚫                       | 🚫         | ✅         | ✅              |
+| Less need to fight dependency arrays                               | 🚫                       | 🚫         | ✅         | ✅              |
+| Easy to share state for sibling components                         | 🚫                       | 🚫         | ✅         | ✅              |
+| Pre-wired to Coder auth logic                                      | 🚫                       | 🚫         | 🚫         | ✅              |
+| Can consume Coder API directly from query function                 | 🚫                       | 🚫         | 🚫         | ✅              |
+| Automatically groups Coder-related queries by prefixing query keys | 🚫                       | 🚫         | 🚫         | ✅              |
+
```
Comment (Kira-Pilot) [inline plugins/backstage-plugin-coder/docs/guides/coder-api.md:60]: In line with my comment above, I don't know that we need the `useState` + `useEffect` column.

File: plugins/backstage-plugin-coder/docs/guides/coder-api.md [hunk 26]
```diff
@@ -0,0 +1,257 @@
...
+#### React Query hooks
+
+The Coder plugin uses [React Query/TanStack Query v4](https://tanstack.com/query/v4/docs/framework/react/overview) for all of its data caching. We recommend that you use it for your own data caching, because of the sheer amount of headaches it can spare you.
+
+There are three main hooks that you will likely need:
+
+- `useQuery` - Query and cache data
+- `useMutation` - Perform mutations on an API resource
+- `useQueryClient` - Coordinate queries and mutations
```
Comment (Kira-Pilot) [inline plugins/backstage-plugin-coder/docs/guides/coder-api.md:25]: Seems like basic use cases for fetching don't need these RQ hooks but they're the first thing we talk about.
WDYT about moving your example code snippets up to the top here alongside your hook explanations so users can immediately see how they might use what you've built?

This *could* look something like:
```
### Available hooks
We've exposed a few hooks to get you started building your own Coder components: 

- `useCoderApi` - Exposes an object with all available Coder API methods. For the most part, there is no exposed state on this object; you can consider it a "function bucket".
- `useCoderAuth` - Provides methods and state values for interacting with your current Coder auth session from within Backstage.
- `useCoderQuery` - a convenience hook that simplifies wiring up `useQuery`, `useCoderApi`, and `useCoderAuth`.

// Query Example:


// Mutation Example:

Note that we use [React Query/TanStack Query v4](https://tanstack.com/query/v4/docs/framework/react/overview) in the above examples. When querying, RQ's `useQuery` hook is also a good escape hatch if `useCoderQuery` doesn't meet your needs, but it requires more work to wire up correctly. 
The plugin exposes a `CODER_QUERY_KEY_PREFIX` constant that you can use to group all Coder queries together; read more here(link to your query constant section).

File: plugins/backstage-plugin-coder/docs/guides/coder-api.md [hunk 27]
```diff
@@ -0,0 +1,257 @@
...
+
+All API calls to **any** of the Coder API functions will fail if you have not authenticated yet. Authentication can be handled via any of the official Coder components that can be imported via the plugin. However, if there are no Coder components on the screen, the `CoderProvider` component will automatically\* inject a fallback auth button for letting the user add their auth info.
+
+<-- Add video of auth flow with fallback button -->
+
+Once the user has been authenticated, all Coder API functions will become available. When the user unlinks their auth token (effectively logging out), all cached queries that start with `CODER_QUERY_KEY_PREFIX` will automatically be vacated.
+
+\* This behavior can be disabled. Please see our [advanced API guide](./coder-api-advanced.md) for more information.
+
+## Connecting a custom query client to the Coder plugin
```
Comment (Kira-Pilot) [inline plugins/backstage-plugin-coder/docs/guides/coder-api.md:71]: Should this section be moved to advanced use cases?

File: plugins/backstage-plugin-coder/docs/guides/coder-api.md [hunk 28]
```diff
@@ -0,0 +1,257 @@
...
+
+```tsx
+const customQueryClient = new QueryClient();
+
+<CoderProvider queryClient={customQueryClient}>
+  <YourCustomComponentsThatNeedAccessToTheCoderPlugin />
+</CoderProvider>;
+```
+
+### Grouping queries with the Coder query key prefix
```
Comment (Kira-Pilot) [inline plugins/backstage-plugin-coder/docs/guides/coder-api.md:92]: Should we move this up so its closer to the hook section

File: plugins/backstage-plugin-coder/docs/guides/coder-api.md [hunk 29]
```diff
@@ -0,0 +1,257 @@
...
+
+const workspacesQuery2 = useCoderQuery({
+  // The query key will automatically have CODER_QUERY_KEY_PREFIX added to the
+  // beginning
+  queryKey: ['workspaces'],
+  queryFn: ({ coderApi }) =>
+    coderApi.getWorkspaces({
+      limit: 10,
+    }),
+});
```
Comment (Kira-Pilot) [inline plugins/backstage-plugin-coder/docs/guides/coder-api.md:115]: I think the first two examples here are unnecessary given the preceding paragraph.
The next two examples are helpful in defining what 'group' means in this context.

File: plugins/backstage-plugin-coder/docs/guides/coder-api.md (hunk 24 above)
Comment (Kira-Pilot) [reply plugins/backstage-plugin-coder/docs/guides/coder-api.md:43]: That makes sense! Could we just move it after the meat of the examples then?

Comment (Parkreiner) [reply plugins/backstage-plugin-coder/docs/guides/coder-api.md:43]: Honestly, I would expect a good chunk of plugin users not to be super seasoned with React

File: plugins/backstage-plugin-coder/docs/guides/coder-api.md (hunk 26 above)
Comment (Parkreiner) [reply plugins/backstage-plugin-coder/docs/guides/coder-api.md]: Yeah, I see what you mean. I went ahead and reorganized the hooks section, and added some bite-sized examples

**Comment by Parkreiner** [comment]
@Kira-Pilot [Third time's the charm](https://github.com/coder/backstage-plugins/blob/mes/vendored-sdk-readme/plugins/backstage-plugin-coder/docs/guides/coder-api.md)?

File: plugins/backstage-plugin-coder/docs/README.md [hunk 30]
```diff
@@ -1,11 +1,18 @@
 # Plugin API Reference – Coder for Backstage
 
-For users who need more information about how to extend and modify the Coder plugin. For general setup, please see our main [README](../README.md).
+The Coder plugin for Backstage does follow semantic versioning.
```
Comment (bpmct) [inline plugins/backstage-plugin-coder/docs/README.md:3]: I think this is an important sentence, but should not necessarily be the first sentence of this doc. Can we change it to something that explains what the reader will learn from reading this doc? e.g.

> Learn how to build custom Backstage functions and components that use the Coder API

**Review by Kira-Pilot — APPROVED** [review]
This looks great. I can see everything I need right up top and there's a great level of detail if I dig in.
